pkg/util/active_groups.go

```diff
@@ -0,0 +1,142 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package util
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/grafana/dskit/services"
+	"go.uber.org/atomic"
+)
+
+const maxGroupsPerUser int = 1000
+
+type ActiveGroups struct {
+	mu                sync.RWMutex
+	timestampsPerUser map[string]map[string]*atomic.Int64 // map[user][group] -> timestamp
+}
+
+func NewActiveGroups() *ActiveGroups {
+	return &ActiveGroups{
+		timestampsPerUser: map[string]map[string]*atomic.Int64{},
+	}
+}
+
+func (ag *ActiveGroups) UpdateGroupTimestampForUser(userID, group string, ts int64) {
+	// Create new atomic before lock is acquired
+	newAtomic := atomic.NewInt64(ts)
+
+	ag.mu.RLock()
+	groupTimestamps := ag.timestampsPerUser[userID]
+	ag.mu.RUnlock()
+
+	if groupTimestamps == nil {
+		ag.mu.Lock()
+		ag.timestampsPerUser[userID] = make(map[string]*atomic.Int64)
```

Comment:
You cannot assume that `ag.timestampsPerUser[userID]` is nil here, as there was a window where you were not holding a write lock. Should be checked again.